src/backend/distributed/utils/citus_clauses.c

```diff
@@ -39,6 +41,146 @@ static Node * PartiallyEvaluateExpressionMutator(Node *expression,
+}
+
+
+static bool
+contain_volatile_functions_checker_citus(Oid func_id, void *context)
+{
+	if (func_id == CitusReadIntermediateResultFuncId())
+	{
+		return false;
+	}
+
+	return (func_volatile(func_id) == PROVOLATILE_VOLATILE);
+}
+
+
+static bool
+contain_volatile_functions_walker_citus(Node *node, void *context)
+{
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	/* Check for volatile functions in node itself */
+	if (check_functions_in_node(node, contain_volatile_functions_checker_citus,
+								context))
+	{
+		return true;
+	}
+
+#if (PG_VERSION_NUM >= 100000)
+	if (IsA(node, NextValueExpr))
+	{
+		/* NextValueExpr is volatile */
+		return true;
+	}
+#endif
+
+	/*
+	 * See notes in contain_mutable_functions_walker about why we treat
+	 * MinMaxExpr, XmlExpr, and CoerceToDomain as immutable, while
+	 * SQLValueFunction is stable.  Hence, none of them are of interest here.
+	 */
+
+	/* Recurse to check arguments */
+	if (IsA(node, Query))
+	{
+		/* Recurse into subselects */
+		return query_tree_walker((Query *) node,
+								 contain_volatile_functions_walker_citus,
+								 context, 0);
+	}
+	return expression_tree_walker(node, contain_volatile_functions_walker_citus,
+								  context);
+}
+
+
+static bool
+contain_mutable_functions_citus(Node *clause)
+{
+	return contain_mutable_functions_walker_citus(clause, NULL);
+}
+
+
+static bool
+contain_mutable_functions_checker_citus(Oid func_id, void *context)
+{
+	if (func_id == CitusReadIntermediateResultFuncId())
+	{
+		return false;
+	}
+	else
+	{
+		return (func_volatile(func_id) != PROVOLATILE_IMMUTABLE);
+	}
+}
+
+
+static bool
+contain_mutable_functions_walker_citus(Node *node, void *context)
+{
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	/* Check for mutable functions in node itself */
+	if (check_functions_in_node(node, contain_mutable_functions_checker_citus,
+								context))
+	{
+		return true;
+	}
+
+#if (PG_VERSION_NUM >= 100000)
+	if (IsA(node, SQLValueFunction))
+	{
+		/* all variants of SQLValueFunction are stable */
+		return true;
+	}
+
+	if (IsA(node, NextValueExpr))
+	{
+		/* NextValueExpr is volatile */
+		return true;
+	}
+#endif
+
+	/*
+	 * It should be safe to treat MinMaxExpr as immutable, because it will
+	 * depend on a non-cross-type btree comparison function, and those should
+	 * always be immutable.  Treating XmlExpr as immutable is more dubious,
+	 * and treating CoerceToDomain as immutable is outright dangerous.  But we
+	 * have done so historically, and changing this would probably cause more
+	 * problems than it would fix.  In practice, if you have a non-immutable
+	 * domain constraint you are in for pain anyhow.
+	 */
+
+	/* Recurse to check arguments */
+	if (IsA(node, Query))
+	{
+		/* Recurse into subselects */
+		return query_tree_walker((Query *) node,
+								 contain_mutable_functions_walker_citus,
+								 context, 0);
+	}
+	return expression_tree_walker(node, contain_mutable_functions_walker_citus,
```

Comment:
Do we need a new walker or could we use `FindNodeCheck`?

Reply:
I don't follow. We want the logic in `contain_mutable_functions` but we don't want our custom functions to be counted for that. We also need the logic `check_functions_in_node` which we don't have in `FindNodeCheck`.

Reply:
It doesn't make much sense that every function has the same expression_walker boilerplate, even (/especially) if we're copy-pasting code from postgres.

For example, you could have this function:
```
static bool
CitusContainsMutableFunction(Node *node)
{
	/* Check for volatile functions in node itself */
	if (check_functions_in_node(node, contain_volatile_functions_checker_citus,
								context))
	{
		return true;
	}

#if (PG_VERSION_NUM >= 100000)
	if (IsA(node, NextValueExpr))
	{
		/* NextValueExpr is volatile */
		return true;
	}
#endif

	return false;
}
```

and then call `FindNodeCheck(node, CitusContainsMutableFunction)`.